File: day7_part1_2015.rs

use std::collections::HashMap;
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Error reading from file");
    println!("{}", some_assembly_required(&input));
}

fn some_assembly_required(input: &str) -> u16 {
    let mut wire_to_rule: HashMap<String, String> = HashMap::new();

    for inst in input.trim().split("\n") {
        let parts: Vec<&str> = inst.split(" -> ").collect();
        wire_to_rule.insert(parts[1].to_string(), parts[0].to_string());
    }

    memo_dfs(&wire_to_rule, "a", &mut HashMap::new())
}

fn memo_dfs(graph: &HashMap<String, String>, entry: &str, memo: &mut HashMap<String, u16>) -> u16 {
    if let Some(&memo_val) = memo.get(entry) {
        return memo_val;
    }

    if entry.chars().all(char::is_numeric) {
        return entry.parse().unwrap();
    }

    let source_rule = graph.get(entry).unwrap();
    let parts: Vec<&str> = source_rule.split(" ").collect();

    let result: u16;
    match parts.len() {
        1 => result = memo_dfs(graph, parts[0], memo),
        2 if parts[0] == "NOT" => {
            let start = memo_dfs(graph, parts[1], memo);
            result = !start;
        }
        3 => match parts[1] {
            "AND" => result = memo_dfs(graph, parts[0], memo) & memo_dfs(graph, parts[2], memo),
            "OR" => result = memo_dfs(graph, parts[0], memo) | memo_dfs(graph, parts[2], memo),
            "LSHIFT" => result = memo_dfs(graph, parts[0], memo) << memo_dfs(graph, parts[2], memo),
            "RSHIFT" => result = memo_dfs(graph, parts[0], memo) >> memo_dfs(graph, parts[2], memo),
            _ => panic!("Invalid operation"),
        },
        _ => panic!("Invalid rule"),
    }

    memo.insert(entry.to_string(), result);
    result
}

File: objectivec/day7_part1_2015.m
#import <Foundation/Foundation.h>

int memoDFS(NSDictionary *graph, NSString *entry, NSMutableDictionary *memo) {
    if ([memo objectForKey:entry]) {
        return [[memo objectForKey:entry] intValue];
    }

    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"[0-9]" options:0 error:nil];
    if ([regex numberOfMatchesInString:entry options:0 range:NSMakeRange(0, [entry length])]) {
        return [entry intValue];
    }

    NSString *sourceRule = [graph objectForKey:entry];
    NSArray *parts = [sourceRule componentsSeparatedByString:@" "];

    int result;
    if ([parts count] == 1) {
        result = memoDFS(graph, [parts objectAtIndex:0], memo);
    } else if ([[parts objectAtIndex:0] isEqualToString:@"NOT"]) {
        int start = memoDFS(graph, [parts objectAtIndex:1], memo);
        result = UINT16_MAX ^ start;
    } else if ([[parts objectAtIndex:1] isEqualToString:@"AND"]) {
        result = memoDFS(graph, [parts objectAtIndex:0], memo) & memoDFS(graph, [parts objectAtIndex:2], memo);
    } else if ([[parts objectAtIndex:1] isEqualToString:@"OR"]) {
        result = memoDFS(graph, [parts objectAtIndex:0], memo) | memoDFS(graph, [parts objectAtIndex:2], memo);
    } else if ([[parts objectAtIndex:1] isEqualToString:@"LSHIFT"]) {
        result = memoDFS(graph, [parts objectAtIndex:0], memo) << memoDFS(graph, [parts objectAtIndex:2], memo);
    } else if ([[parts objectAtIndex:1] isEqualToString:@"RSHIFT"]) {
        result = memoDFS(graph, [parts objectAtIndex:0], memo) >> memoDFS(graph, [parts objectAtIndex:2], memo);
    }

    [memo setObject:[NSNumber numberWithInt:result] forKey:entry];
    return result;
}

int main() {
    NSError *error;
    NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:&error];
    if (error) {
        printf("Error reading from file: %s\n", [[error localizedDescription] UTF8String]);
        return 1;
    }

    NSMutableDictionary *wireToRule = [NSMutableDictionary dictionary];
    NSArray *inst = [input componentsSeparatedByString:@"\n"];
    for (NSString *line in inst) {
        NSArray *parts = [line componentsSeparatedByString:@" -> "];
        [wireToRule setObject:[parts objectAtIndex:0] forKey:[parts objectAtIndex:1]];
    }

    int aSignal = memoDFS(wireToRule, @"a", [NSMutableDictionary dictionary]);
    printf("%d\n", aSignal);

    return 0;
}